### PR TITLE
Add ability to server render

### PR DESCRIPTION
**Pull request by chris-kruining**
pertains to #24

### CODE REVIEW

**Comment by chris-kruining** [comment]
DO NOT MERGE YET

I still have an issue with the portal, react seems to loose track of the children, so those are rendered twice with the current code

**Comment by chris-kruining** [comment]
this repo seems to be dead, so I forked instead

**Comment by apearce** [comment]
I'll take a look. I probably won't have time until this weekend.

File: src/lib/ReactShadowRoot.js [hunk 1]
```diff
@@ -23,7 +23,7 @@ export default class extends React.PureComponent {
   static propTypes = {
     delegatesFocus: PropTypes.bool,
     mode: PropTypes.oneOf(['open', 'closed']),
-    stylesheets: PropTypes.arrayOf(PropTypes.instanceOf(window.CSSStyleSheet))
+    stylesheets: PropTypes.arrayOf(PropTypes.instanceOf(typeof window !== 'undefined' && window.CSSStyleSheet))
```
Comment (apearce) [inline src/lib/ReactShadowRoot.js:26]: In order to get it to compile I had to change this to `stylesheets: PropTypes.arrayOf(typeof window !== 'undefined' ? PropTypes.instanceOf(window.CSSStyleSheet) : PropTypes.any)

File: src/lib/ReactShadowRoot.js [hunk 2]
```diff
@@ -47,10 +47,11 @@ export default class extends React.PureComponent {
       stylesheets
     } = this.props;
 
-    this.shadowRoot = this.placeholder.current.parentNode.attachShadow({
-      delegatesFocus,
-      mode
-    });
+    this.shadowRoot = this.placeholder.current.parentNode.shadowRoot
```
Comment (apearce) [inline src/lib/ReactShadowRoot.js:50]: Is this change necessary? I modified the `lib` file in a remix app locally and it worked without it.

File: src/lib/ReactShadowRoot.js [hunk 3]
```diff
@@ -63,7 +64,7 @@ export default class extends React.PureComponent {
 
   render() {
     if (!this.state.initialized) {
-      return <span ref={this.placeholder}></span>;
+      return <template shadowroot="open" ref={this.placeholder}>{this.props.children}</template>;
```
Comment (apearce) [inline src/lib/ReactShadowRoot.js:67]: Can you elaborate on why this is needed? Does it have something to do with Remix?  I modified the `lib` file in a remix app locally and it worked without it.

File: src/lib/ReactShadowRoot.js (hunk 2 above)
Comment (chris-kruining) [reply src/lib/ReactShadowRoot.js:50]: yes, and that has to do with render also rendering the template html, this is how declaritive Shadow-root works: https://web.dev/declarative-shadow-dom/

File: src/lib/ReactShadowRoot.js (hunk 3 above)
Comment (apearce) [reply src/lib/ReactShadowRoot.js:67]: Thanks. I didn't know that existed. Really cool.

Comment (apearce) [reply src/lib/ReactShadowRoot.js:67]: This is causing a problem with the build. Looks like it's something to do with Typescript

Type '{ children: ReactNode; shadowroot: any; ref: RefObject<any>; }' is not assignable to type 'DetailedHTMLProps<HTMLAttributes<HTMLTemplateElement>, HTMLTemplateElement>'.
  Property 'shadowroot' does not exist on type 'DetailedHTMLProps<HTMLAttributes<HTMLTemplateElement>, HTMLTemplateElement>'.

Comment (apearce) [reply src/lib/ReactShadowRoot.js:67]: I had to `// @ts-ignore` this line to get it to build. I also changed the `shadowroot` attribute to take `this.props.mode`.

I need to do some testing on this as I have a few concerns. I am moving tomorrow and will be pretty busy for the next week or so but will try to work on it when I can.

Comment (chris-kruining) [reply src/lib/ReactShadowRoot.js:67]: I also couldn't get the shadowroot attribute to work, it also doesn't help that HTMLTemplateElement is one of those few elements in React that is not easily extended due to it's setup....

Also, no problem, take your time. good luck with moving :D

Comment (apearce) [reply src/lib/ReactShadowRoot.js:67]: I didn't forget about this! Moving and getting settled take longer than I thought.
I'm trying to figure out the best way to handle this since declarative shadow-root basically breaks delegatesFocus and constructable stylesheets. I may make declarative opt-in.

Comment (apearce) [reply src/lib/ReactShadowRoot.js:67]: So what I _think_ is happening with the double rendering is the server still sends back the template tag, the shadow root is applied by the browser. When React hydrates is puts the children in the template tag but they are already there from the react-shadow-root code. Then the browser takes everything in the template tag and puts it in the shadow root of the parent element.

So rather than reusing the shadow root on the template, you have to call attachShadow on it which will replace it and things only render once. With JS turned off the children are still but in the shadow root of the parent by the browser. 

I have this working locally but need to do a little more testing before merging it in.

Comment (chris-kruining) [reply src/lib/ReactShadowRoot.js:67]: I think your logic is wrong, the template is "consumed/converted" during html parsing (as in, in the parse stream itself, before any js is run, afaik). My suspicion is that it is not to much that react is rendering the children twice, that is to say, that is a symptom, not the cause. I Recon react loses where the children are because the browser changes the DOM in a way react cannot handle. so when it tries to render / sync the v-dom it concludes the children are to be added.

Then again, if you seem to have solved it and the PR is almost ready: ❤❤❤❤

Comment (apearce) [reply src/lib/ReactShadowRoot.js:67]: I pushed out a new version that supports declarative shadow root. 6.2.0 is the version. You have to set the `declarative` prop to `true` to use it.

See https://apearce.github.io/react-shadow-root/#declarative for additional info.

Thanks for bringing this to my attention. I had no idea it existed.
